future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,569 @@
+     */
+    function getEmployee(uint256 employeeId)
+        public
+        view
+        returns(
+            address accountAddress,
+            uint256 yearlyDenominationSalary,
+            string name,
+            uint lastAllocation,
+            uint lastPayroll
+        )
+    {
+        var employee = employees[employeeId];
+
+        accountAddress = employee.accountAddress;
+        yearlyDenominationSalary = employee.denominationTokenSalary.toYearlyDenomination();
+        name = employee.name;
+        lastAllocation = employee.lastAllocation;
+        lastPayroll = employee.lastPayroll;
+    }
+
+    /**
+     * @dev Get total amount of salaries in Payroll
+     * @notice Get total amount of salaries in Payroll
+     * @return Integer with the amount
+     */
+    function getYearlyTotalPayroll() public view returns(uint256 total) {
+        total = yearlyTotalPayroll;
+    }
+
+    /**
+     * @dev Monthly Denomination amount spent in salaries
+     * @notice Monthly Denomination amount spent in salaries
+     * @return Integer with the monthly amount
+     */
+    function calculatePayrollBurnrate() public view returns(uint256 payrollBurnrate) {
+        payrollBurnrate = yearlyTotalPayroll / 12;
+    }
+
+    /**
+     * @dev Days until the contract can run out of funds
+     * @notice Days until the contract can run out of funds
+     * @return Integer with the number of days
+     */
+    function calculatePayrollRunway() public view returns(uint256 payrollRunway) {
+        if (yearlyTotalPayroll == 0)
+            return MAX_UINT;
+
+        uint256 balance = 0;
+        // for each token, get finance balance and convert it to Denomination
+        for (uint i = 0; i < allowedTokensArray.length; i ++) {
+            var token = allowedTokensArray[i];
+            var tokenContract = ERC20(token);
+            if (exchangeRates[token] == 0)
+                continue;
+            balance = balance.add(tokenContract.balanceOf(finance).div(exchangeRates[token]));
+        }
+        payrollRunway = balance.mul(365) / yearlyTotalPayroll;
+    }
+
+    /**
+     * @dev Set token distribution for payments to an employee (the caller)
+     * @notice Set token distribution for payments to an employee (the caller).
+     *         Only callable once every 6 months
+     * @param tokens Array with the tokens to receive, they must belong to allowed tokens for employee
+     * @param distribution Array (correlated to tokens) with the proportions (integers over 100)
+     */
+    function determineAllocation(address[] tokens, uint256[] distribution) public {
+        var employee = employees[employeeIds[msg.sender]];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
+        // check that enough time has gone by
+        require(getTimestamp() > employee.lastAllocation && getTimestamp() - employee.lastAllocation > 15768000); // half a year in seconds
```

Comment:
lets remove the last allocation time

Reply:
So, you mean allowing employees to change allocation whenever they want?